Sprint #16 homework

ekalemen opened the pull request.

Commit:
Sprint#16: migrate to ViewBinding

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>

Commit:
Sprint#16: refactor project structure

This refactors project structure from arch-layers->features to features->arch-layers.

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>

Commit:
Sprint#16: refator project using MVVM arch template

This refactor Settings, Player, and Track Search screens using
MVVM template.

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/main/ui/MainActivity.kt:
```diff
@@ -1,34 +1,35 @@
+        setContentView(binding.root)
 
-        val buttonMainSearch = findViewById<Button>(R.id.main_butt_search)
+        /*val buttonMainSearch = findViewById<Button>(R.id.main_butt_search)
```
Закоментированный код лучше сразу удалять. Хранить закоментированный код в проекте - плохая практика, в случае чего этот код можно восстановить с помощью истории гита

ekalemen replied:
fixed

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/player/ui/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,104 @@
+        viewModel.playerPlayingTime.observe(this) { time -> renderTrackTime(time) }
+
+        val intent = intent
+        val trackStr = intent.getStringExtra(EXTRA_TRACK_INFO)
```
Для передачи данных между компонентами Android (активити, фрагменты, сервисы) лучше использовать интерфейс `Parcelable`. Для генерации реализации этого интерфейса есть плагин `kotlin-parcelize`. Чтобы реализовать нужно сделать следующее - добавить плагин в проект, доменную модель трека наследовать от интерфейса `Parcelable` и воспользоваться аннотацией `@Parcelize`. Ну и передавать сами данные не в формате String, а сразу в формате Parcelable. Тут есть пример работы с плагином - https://developer.android.com/kotlin/parcelize

ekalemen replied:
fixed

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/player/ui/viewmodel/PlayerViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+            track.previewUrl,
+            {
+                _playerStatus.postValue(PlayerStatus.STATE_PREPARED)
+                _playerPlayingTime.postValue("00:00")
```
Лучше не хардкодить этот текст, так как форматирование может легко поменяться и потом придётся по всему проекту искать эти захардкоженные строчки. Можно либо вынести в strings.xml, чтобы это было только в одном месте, либо использовать ту же функцию форматирования через `SimpleDateFormat`, передавая туда значение 0, чтобы поведение было консистентным

ekalemen replied:
fixed

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/search/ui/viewmodel/SearchViewModel.kt:
```diff
@@ -0,0 +1,128 @@
+    fun onClearTrackHistory() {
+        searchHistoryInteractor.clearTrackHistory()
+        _historyTracks.postValue(listOf())
+        _searchTracks.postValue(listOf())
```
Лучше использовать `emptyList`, так как `listOf()` создаст список, заранее выделив память под 10 элементов

ekalemen replied:
Fixed

ArturNurtdinov reviewed (commented):
1. Элементы истории поиска отображаются при пустой истории поиска:

<img width="533" height="518" alt="image" src="https://github.com/user-attachments/assets/3ee8e176-835f-4a90-b588-ac1f5e34205e" />

2. Плейсхолдер "Ничего не нашлось" отображается над списком треков:

<img width="800" height="474" alt="image" src="https://github.com/user-attachments/assets/40588ffd-2482-40f7-b096-00bda56055e7" />

Commit:
Sprint #16: review fixes

This fixes :
 - use Parcelable instead of Gson converting
 - fix displaying of history view elements on empty history
 - fix displaying of empty placeholder text after successefull search
 - get rid of hardcoded "00:00" starting time
 - get rid of commented code

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>

Commit:
Sprint #16: review fixes #2

This fixes:
 - PlayerInteractorImpl dependency from MediaPlayer
 - ThemeSwitcherInteractorImpl dependency from AppCompatDelegate

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>

Commit:
Sprint#16: review fixes #3

This fixes:
 - null ppointer exception on get saved theme
 - progress bar view state - set placeholder text to empty string

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>
